src/target/riscv/riscv-013.c

```diff
@@ -2811,6 +2815,162 @@ void riscv013_fill_dmi_nop_u64(struct target *target, char *buf)
+      }
+    }
+  }
+  LOG_INFO("System Bus Access Test: Read/write test, no addr autoincrement PASSED");
+
+  // Test 2: Simple write/read test, with address autoincrement 
+  sbcs = set_field(sbcs_orig,DMI_SBCS_SBAUTOINCREMENT,1);
+  dmi_write(target,DMI_SBCS,sbcs);
+
+  for(int sbaccess = 0; sbaccess < 0x5; sbaccess++){
+    sbcs = set_field(sbcs_orig,DMI_SBCS_SBACCESS,sbaccess);
+    dmi_write(target,DMI_SBCS,sbcs);
+
+    dmi_write(target,DMI_SBADDRESS0,0x80000000);
+    for(int i = 0; i < 100; i++){
+      read_sbcs_nonbusy(target,&sbcs);
+      dmi_write(target,DMI_SBDATA0,i);
+    }
+
+    sbcs = set_field(sbcs_orig,DMI_SBCS_SBREADONDATA,1);
+
+    dmi_write(target,DMI_SBADDRESS0,0x80000000);
+    for(uint32_t i = 0; i < 100; i++){
+      read_sbcs_nonbusy(target,&sbcs);
+      uint32_t val;
+      dmi_read(target,&val,DMI_SBDATA0);
+      if(i != val) {
+        LOG_ERROR("System Bus Access Test: Error reading autoincremented address, expected val = %d, read val = %d",i,val);
+      }
+    }
+  } 
+  LOG_INFO("System Bus Access Test: Read/write test, with addr autoincrement PASSED");
+
+  // Test 3: Read from illegal address
+  read_memory_sba_simple(target,address,sbcs_orig);
+
+  dmi_read(target,&rd_val,DMI_SBCS);
+  if(get_field(rd_val,DMI_SBCS_SBERROR) == 0x2) {
+    LOG_INFO("System Bus Access Test: Illegal address read test PASSED");
+    sbcs = set_field(sbcs_orig,DMI_SBCS_SBERROR,1);
+    dmi_write(target,DMI_SBCS,sbcs);
+  } else {
+    LOG_ERROR("System Bus Access Test: Illegal address read test FAILED");
+  }
+
+  // Test 4: Write to illegal address
+  write_memory_sba_simple(target,address,0xdeadbeef,sbcs_orig);
+
+  dmi_read(target,&rd_val,DMI_SBCS);
+  if(get_field(rd_val,DMI_SBCS_SBERROR) == 0x2) {
+    LOG_INFO("System Bus Access Test: Illegal address write test PASSED");
+    sbcs = set_field(sbcs_orig,DMI_SBCS_SBERROR,1);
+    dmi_write(target,DMI_SBCS,sbcs);
+  } else {
+    LOG_ERROR("System Bus Access Test: Illegal address write test FAILED");
+  }
+
+  // Test 5: Write to unsupported sbaccess size 
+  uint32_t sbaccess128 = get_field(sbcs_orig,DMI_SBCS_SBACCESS128);
+
+  if(sbaccess128) {
+    LOG_INFO("System Bus Access Test: SBCS Alignment error test PASSED, all alignments supported");
+  } else {
+    sbcs = set_field(sbcs_orig,DMI_SBCS_SBACCESS,0x4);
+    dmi_write(target, DMI_SBCS, sbcs);
+
+    write_memory_sba_simple(target, 0x80000000, 0xdeadbeef, sbcs_orig);
+
+    dmi_read(target,&rd_val,DMI_SBCS);
+
+    if(get_field(rd_val,DMI_SBCS_SBERROR) == 0x3) {
+      LOG_INFO("System Bus Access Test: SBCS Alignment error test PASSED");
+      sbcs = set_field(sbcs_orig,DMI_SBCS_SBERROR,0x1);
+      dmi_write(target,DMI_SBCS,sbcs);
+    } else {
+      LOG_ERROR("System Bus Access Test: SBCS Alignment error test FAILED");
+    }
+  }
+   
+  // Test 5: Set sbbusyerror
+  dmi_write(target,DMI_SBADDRESS0,0x80000000);
+  dmi_write(target,DMI_SBADDRESS0,0x80000000);
```

Comment:
This isn’t guaranteed to give you a busy error. Maybe you’d have to batch up writes and even that may not be fast enough (JTAG is slow!)

Reply:
Yeah, I think this flag may be impossible to hit unless we add some sim-only logic to increase TileLink's read/write response delay